core/src/main/scala/kafka/server/MetadataCache.scala

```diff
@@ -47,55 +47,79 @@ private[server] class MetadataCache(brokerId: Int) extends Logging {
 
   this.logIdent = "[Kafka Metadata Cache on broker %d] ".format(brokerId)
 
-  def getTopicMetadata(topics: Set[String], protocol: SecurityProtocol) = {
+  private def getAliveEndpoints(brokers: Iterable[Int], protocol: SecurityProtocol): List[Node] = {
+    brokers.flatMap(aliveBrokers.get).toSeq.map(_.getNode(protocol)).toList
```

Comment:
Also, you probably want to move the `toSeq` before the `flatMap` as we don't need set semantics here (in case `brokers` is a `Set`, which it is in the caller).